File: src/java/com/skocur/imagecipher/encrypters/Encrypter.java
package com.skocur.imagecipher.encrypters;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;

public abstract class Encrypter {

    private int imageWidth;
    private int imageHeight;
    protected String fileName;
    protected BufferedImage image;

    public Encrypter(String fileName) throws IOException {
        this.fileName = fileName;
        this.image = ImageIO.read(new File(fileName));
        this.imageWidth = image.getWidth();
    }

    public abstract void encrypt(String text);

    protected void saveEncryptedData() {
        try {
            File file = new File(fileName);
            ImageIO.write(image, "png", file);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public int getImageWidth() {
        return imageWidth;
    }

    public int getImageHeight() {
        return imageHeight;
    }
}

File: app/src/main/kotlin/com/imagecipher/app/encrypters/Encrypter.kt
package com.imagecipher.app.encrypters

import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import javax.imageio.ImageIO

/**
 * Abstract class representing encrypters within Image Cipher. Every encrypter operates only on pixels of
 * a loaded image. Output is always saved as soon as encryption is done. Getter methods that return
 * dimensions of the image can be used by outside classes to calculate the capacity of data that are allowed
 * to be stored.
 */
abstract class Encrypter(protected var fileName: String) : AutoCloseable {

    protected var image: BufferedImage = ImageIO.read(File(fileName))
    private val imageWidth: Int = image.width

    /**
     * Abstract method for encrypting text.
     *
     * @param text Text to be encrypted
     */
    abstract fun encrypt(text: String)

    /**
     * Method that should be invoked after all encryption processes have finished.
     */
    override fun close() {
        try {
            val file = File(fileName)
            ImageIO.write(image, "png", file)
        } catch (e: IOException) {
            e.printStackTrace()
        }
    }

    fun getImageWidth(): Int {
        return imageWidth
    }
}
